iree/compiler/Dialect/Stream/Transforms/MaterializeBuiltins.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <utility>
+
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamDialect.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Stream/IR/StreamTypes.h"
+#include "iree/compiler/Dialect/Stream/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+
+// We could overengineer this with custom DSLs and python generation and all
+// that kind of stuff - but that stuff really belongs closer to the frontend
+// (linalg/etc). Once we are at this point we are just patching over things for
+// compatibility and not handling arbitrary programs. A linalg.fill of an i64
+// that gets tiled and fused with other operations is always going to be several
+// of orders of magnitude faster than this approach and we should spend our
+// effort improving things at that layer instead of leaning too much on this.
```

Comment:
(Just thinking aloud here, I don't think specific actions are needed)

It could be nice to call this out somehow for frontend authors, but ideally this would just show up in traces and profiles if programs do end up using this in ways that would benefit from optimization or more generality. This doesn't seem like a big architectural flaw either - frontends should be capable of setting up programs more efficiently and this has no cost (beyond code complexity in the compiler) to programs that don't need it.

Reply:
yeah - we'll at least see the "builtin" dispatches named nicely and can see when they're taking time - it's kind of like when a memset/memcpy gets inlined and goes away vs you see a full call when profiling cpu code